File: majority-element.py
class Solution:
    # sort, fast
    def majorityElement(self, nums: List[int]) -> int:
        return sorted(nums)[len(nums) // 2]
    
    # divide and conquer
    def majorityElement(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        
        half = len(nums) // 2
        a = self.majorityElement(nums[:half])
        b = self.majorityElement(nums[half:])
        return [a, b][nums.count(b) > half]

    # collections.Counter
    def majorityElement(self, nums: List[int]) -> int:
        num, count = collections.Counter(nums).most_common(1)[0]
        if count > len(nums) // 2:
            return num
        else:
            return

    # collections.Counter (under condition that the majority element always exists)
    def majorityElement(self, nums: List[int]) -> int:
        return collections.Counter(nums).most_common(1)[0][0]
    
    # dynamic programming (memoization)
    def majorityElement(self, nums: List[int]) -> int:
        count = collections.defaultdict(int)
        for num in nums:
            if count[num] == 0:
                count[num] = nums.count(num)
            if count[num] > len(nums) // 2:
                return num

File: majority-element.java
class Solution {
    public int majorityElement(int[] nums) {
        Arrays.sort(nums);
        return nums[nums.length/2];
    }
}

class Solution {
    public int majorityElement(int[] nums) {
        return mergeSort(nums)[nums.length / 2];
    }

    public int[] mergeSort(int[] nums) {
        if(nums.length == 1)
            return nums;
        
        int[] left = mergeSort(Arrays.copyOfRange(nums, 0, nums.length / 2));
        int[] right = mergeSort(Arrays.copyOfRange(nums, nums.length / 2, nums.length));

        int[] sorted = new int[nums.length];
        int index = 0, x = 0, y = 0;
        while(x < left.length && y < right.length) {
            if(left[x] < right[y]) {
                sorted[index++] = left[x++];
            } else {
                sorted[index++] = right[y++];
            }
        }
        while(x < left.length)
            sorted[index++] = left[x++];
        while(y < right.length)
            sorted[index++] = right[y++];

        return sorted;
    }
}